tests/slow/src/test/scala/org/scalafmt/internal/PropertyTest.scala

```diff
@@ -0,0 +1,123 @@
+package org.scalafmt.internal
+
+import java.io.File
+import java.nio.charset.StandardCharsets
+import java.nio.file._
+import java.util.concurrent.atomic.AtomicInteger
+
+import me.tongfei.progressbar.{ProgressBar => PB, ProgressBarStyle}
+
+import org.scalafmt.tests.BaseScalaPrinterTest
+import org.scalameta.logger
+
+import scala.meta._
+import scala.meta.testkit.Corpus
+
+import scala.collection.concurrent.TrieMap
+import scala.util.control.NonFatal
+
+abstract class PropertyTest(name: String) extends BaseScalaPrinterTest {
+
+  val failed = TrieMap.empty[File, Boolean]
+  val regressions = TrieMap.empty[File, Boolean]
+  val nl = "\n"
+  val prefix = "target/repos/"
+
+  val coverageFile = Paths.get(s"coverage-${name}.txt")
+  val todoFile = Paths.get(s"todo-${name}.diff")
+
+  if (Files.exists(todoFile)) {
+    Files.delete(todoFile)
+  }
+
+  val previouslyFailed: Set[File] =
+    if (Files.exists(coverageFile)) {
+      val input = new String(Files.readAllBytes(coverageFile))
+      input.split(nl).filterNot(_ == "").map(f => new File(prefix + f)).toSet
+    } else {
+      Set()
+    }
+
+  def check(file: Input.File): Either[String, Unit]
```

Comment:
Can we move the abstract method to the top and add a comment explaining what `Either[String, Unit]`  means